main/src/ca/uwaterloo/flix/api/lsp/provider/CompletionProvider.scala

```diff
@@ -0,0 +1,454 @@
+      kind = CompletionItemKind.Function)
+  }
+
+  private def sigCompletion(decl: TypedAst.Sig)(implicit context: Context, index: Index, root: TypedAst.Root): CompletionItem = {
+    val name = decl.sym.toString
+    CompletionItem(label = getLabelForNameAndSpec(decl.sym.toString, decl.spec),
+      sortText = "2" + name,
+      textEdit = TextEdit(context.range, getApplySnippet(name, decl.spec.fparams)),
+      detail = Some(FormatScheme.formatScheme(decl.spec.declaredScheme)),
+      documentation = Some(decl.spec.doc.text),
+      insertTextFormat = InsertTextFormat.Snippet,
+      kind = CompletionItemKind.Interface)
+  }
+
+  /**
+    * Returns `true` if the given definition `decl` should be included in the suggestions.
+    */
+  private def matchesDef(decl: TypedAst.Def, word: String, uri: String): Boolean = {
+    val isPublic = decl.spec.mod.isPublic
+    val isNamespace = word.nonEmpty && word.head.isUpper
+    val isMatch = if (isNamespace)
+                    decl.sym.toString.startsWith(word)
+                  else
+                    decl.sym.text.startsWith(word)
+    val isInFile = decl.sym.loc.source.name == uri
+
+    isMatch && (isPublic || isInFile)
+  }
+
+  /**
+    * Returns `true` if the given signature `sign` should be included in the suggestions.
+    */
+  private def matchesSig(sign: TypedAst.Sig, word: String, uri: String): Boolean = {
+    val isPublic = sign.spec.mod.isPublic
+    val isNamespace = word.nonEmpty && word.head.isUpper
+    val isMatch = if (isNamespace)
+                    sign.sym.toString.startsWith(word)
+                  else
+                    sign.sym.name.startsWith(word)
+    val isInFile = sign.sym.loc.source.name == uri
+
+    isMatch && (isPublic || isInFile)
+  }
+
+  private def getDefAndSigCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): Iterable[CompletionItem] = {
+    if (root == null) {
+      return Nil
+    }
+
+    val word = context.word
+    val uri = context.uri
+
+    val defSuggestions = root.defs.values.filter(matchesDef(_, word, uri)).map(defCompletion)
+    val sigSuggestions = root.sigs.values.filter(matchesSig(_, word, uri)).map(sigCompletion)
+    defSuggestions ++ sigSuggestions
+  }
+
+  /**
+    * Returns a list of completion items based on with type class constraints.
+    */
+  private def getWithCompletions()(implicit context: Context, index: Index, root: TypedAst.Root): Iterable[CompletionItem] = {
+    if (root == null) {
+      return Nil
+    }
+
+    // 
+    // When used with `enum`, `with` needs to be treated differently: we should only show derivable
+    // type classes, and we shouldn't include the type parameter
+    // 
+
+    val enumPattern = raw"\s*enum\s+(.*\s)wi?t?h?\s?.*".r
+    val withPattern = raw"\s*(def|instance|class)\s+(.*\s)wi?t?h?\s?.*".r
+    val wordPattern = "wi?t?h?".r
+
+    val currentWordIsWith = wordPattern matches context.word
+
+    if (enumPattern matches context.prefix) {
```

Comment:
I think in the future, but perhaps not for this PR, we should refactor each branch into its own helper function.

Reply:
I guess they're different enough that the commonality between them isn't worth retaining a single function. But I agree, I'll leave it like this for now.